src/Linux/Host/Broker/Impl/Services/Utility.cs

```diff
@@ -1,61 +1,98 @@
-﻿
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
 using System;
-using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
 using System.Reflection;
 using System.Text;
 using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
 using Microsoft.Common.Core.OS;
 using Microsoft.Extensions.Logging;
 using Microsoft.R.Host.Broker.Startup;
 using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
 
 namespace Microsoft.R.Host.Broker.Services {
     public class Utility {
+        private const string PamInfo = "pam-info";
+        private const string PamError = "pam-error";
+        private const string JsonError = "json-error";
+        private const string RtvsResult = "rtvs-result";
+        private const string RtvsError = "rtvs-error";
+
         public static bool AuthenticateUser(ILogger<IAuthenticationService> logger, IProcessServices ps,  string username, string password, out string profileDir) {
             bool retval = false;
             Process proc = null;
             string userDir = string.Empty;
             try {
                 proc = CreateRLaunchProcess(ps, true);
-                using (BinaryWriter writer = new BinaryWriter(proc.StandardInput.BaseStream))
-                using (BinaryReader reader = new BinaryReader(proc.StandardOutput.BaseStream)) {
-                    AuthenticationOnlyMessage message = new AuthenticationOnlyMessage();
-                    message.Username = username;
-                    message.Password = password;
-
-                    string json = JsonConvert.SerializeObject(message);
-                    byte[] jsonBytes = Encoding.UTF8.GetBytes(json);
-
+                using (BinaryWriter writer = new BinaryWriter(proc.StandardInput.BaseStream, Encoding.UTF8, true))
+                using (BinaryReader reader = new BinaryReader(proc.StandardOutput.BaseStream, Encoding.UTF8, true)) {
+                    AuthenticationOnlyMessage message = new AuthenticationOnlyMessage() { Username = username, Password = password };
+                    string json = JsonConvert.SerializeObject(message, GetJsonSettings());
+                    var jsonBytes = Encoding.UTF8.GetBytes(json);
                     writer.Write(jsonBytes.Length);
                     writer.Write(jsonBytes);
                     writer.Flush();
 
-                    // wait for the process to exit;
-                    while (!proc.HasExited) {
-                        Thread.Sleep(100);
-                    }
+                    var waitForExit = Task.Run(() => {
```

Comment:
oops. forgot about that one.

---
In reply to: [116115166](https://github.com/Microsoft/RTVS/pull/3516#discussion_r116115166) [](ancestors = 116115166)